otlp_json2/__init__.py

```diff
@@ -0,0 +1,173 @@
+from __future__ import annotations
+
+import json
+from collections.abc import Mapping, Sequence
+from typing import Any, TYPE_CHECKING
+
+if TYPE_CHECKING:
+    from typing_extensions import TypeAlias
+
+    from opentelemetry.trace import Link
+    from opentelemetry._logs import LogRecord
+    from opentelemetry.sdk.trace import ReadableSpan, Event
+    from opentelemetry.sdk.resources import Resource
+    from opentelemetry.sdk.util.instrumentation import InstrumentationScope
+    from opentelemetry.trace.status import Status
+
+    _LEAF_VALUE: TypeAlias = "str | int | float | bool"  # TODO: confirm
+    _VALUE: TypeAlias = "_LEAF_VALUE | Sequence[_LEAF_VALUE]"
+
+
+CONTENT_TYPE = "application/json"
+
+
+_VALUE_TYPES = {
+    # NOTE: order matters, for isinstance(True, int).
+    bool: ("boolValue", bool),
+    int: ("intValue", str),
+    float: ("doubleValue", float),
+    bytes: ("bytesValue", bytes),
+    str: ("stringValue", str),
+    Sequence: (
+        "arrayValue",
+        lambda value: {"values": [_value(e) for e in _homogeneous_array(value)]},
+    ),
+    Mapping: (
+        "kvlistValue",
+        lambda value: {"values": [{k: _value(v) for k, v in value.items()}]},
+    ),
+}
+
+
+def encode_spans(spans: Sequence[ReadableSpan]) -> bytes:
+    spans = sorted(spans, key=lambda s: (id(s.resource), id(s.instrumentation_scope)))
```

Comment:
Why are you sorting by the `id`? Won't that be pseudo-random, based on the memory address of the CPython object? Should this be `lambda s: (s.resource, s.instrumentation_scope)`?

Reply:
I see -- that seems reasonable. Do those objects have a more human identity/name? In any case, a comment explaining what's going on here would be helpful.

Reply:
Not really...

A Resource (there's typically only one anyway) is just what the SDK was initialised with. It doesn't even have a name, only attributes.

An InstrumentationScope is the result of `tracer = xx.get_tracer(name, version)` where the name could be `__name__` or in our case `"ops"` for the entire ops package, submodules included, while charmers and charm lib authors pick their own names.

Reply:
Chat-chat: If this was not Python, I would probably serialise every span's resource and then group by that serialised string... but here I don't see the point.